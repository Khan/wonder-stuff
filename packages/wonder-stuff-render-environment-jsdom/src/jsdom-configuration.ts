import {KindError, Errors} from "@khanacademy/wonder-stuff-core";
import type {
    RenderAPI,
    ICloseable,
} from "@khanacademy/wonder-stuff-render-server";
import type {IJSDOMConfiguration, CloseableResourceLoader} from "./types";

/**
 * Utility for creating a valid configuration to use with the JSDOM environment.
 */
export class JSDOMConfiguration implements IJSDOMConfiguration {
    readonly registrationCallbackName: string;
    readonly getFileList: (
        url: string,
        renderAPI: RenderAPI,
        fetchFn: (url: string) => Promise<Buffer> | null | undefined,
    ) => Promise<Array<string>>;
    readonly getResourceLoader: (
        url: string,
        renderAPI: RenderAPI,
    ) => CloseableResourceLoader;
    readonly afterEnvSetup: (
        url: string,
        fileURLs: ReadonlyArray<string>,
        renderAPI: RenderAPI,
        vmContext?: any,
    ) => Promise<ICloseable | null | undefined>;

    /**
     * Create a configuration for use with the JSDOM  environment.
     *
     * @param {(url: string, renderAPI: RenderAPI) => Promise<Array<string>>} getFileList
     * Callback that should return a promise for the list of JavaScript files
     * the environment must execute in order to produce a result for the given
     * render request.
     * @param {(url: string, renderAPI: RenderAPI) => ResourceLoader} getResourceLoader
     * Callback that should return a JSDOM resource loader for the given
     * request. We must call this per render so that logging is appropriately
     * channeled for the request being made.
     * @param {(url: string, fileURLs: $ReadOnlyArray<string>, renderAPI: RenderAPI, vmContext: any) => ?Promise<mixed>} [afterEnvSetup]
     * Callback to perform additional environment setup before the render
     * occurs. This can optionally return an object that can add extra fields
     * to the environment context for rendering code to access. This is useful
     * if your render server wants to add some specific configuration, such
     * as setting up some versions of Apollo for server-side rendering.
     * Be careful; any functions you attach can be executed by the rendering
     * code.
     * @param {string} [registrationCallbackName] The name of the function
     * that the environment should expose for client code to register for
     * rendering. This defaults to `__jsdom_env_register`.
     */
    constructor(
        getFileList: (
            url: string,
            renderAPI: RenderAPI,
            fetchFn: (url: string) => Promise<Buffer> | null | undefined,
        ) => Promise<Array<string>>,
        getResourceLoader: (
            url: string,
            renderAPI: RenderAPI,
        ) => CloseableResourceLoader,
        afterEnvSetup?: (
            url: string,
            fileURLs: ReadonlyArray<string>,
            renderAPI: RenderAPI,
            vmContext?: any,
        ) => Promise<ICloseable | null | undefined>,
        registrationCallbackName = "__jsdom_env_register",
    ) {
        if (typeof getFileList !== "function") {
            throw new KindError(
                "Must provide valid callback for obtaining file list",
                Errors.Internal,
            );
        }
        if (typeof getResourceLoader !== "function") {
            throw new KindError(
                "Must provide valid callback for obtaining resource loader",
                Errors.Internal,
            );
        }
        if (afterEnvSetup != null && typeof afterEnvSetup !== "function") {
            throw new KindError(
                "Must provide valid callback for after env setup or null",
                Errors.Internal,
            );
        }

        this.registrationCallbackName = registrationCallbackName;
        this.getFileList = getFileList;
        this.getResourceLoader = getResourceLoader;
        this.afterEnvSetup = afterEnvSetup || (() => Promise.resolve(null));
    }
}
