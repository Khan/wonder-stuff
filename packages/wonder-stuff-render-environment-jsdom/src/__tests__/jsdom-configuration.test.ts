import {JSDOMConfiguration} from "../jsdom-configuration";

describe("JSDOMConfiguration", () => {
    describe("#constructor", () => {
        it.each([null, "not a function"])(
            "should throw if invalid getFileList is provided",
            (badGetFileList: any) => {
                // Arrange

                // Act
                const underTest = () =>
                    new JSDOMConfiguration(badGetFileList, jest.fn());

                // Assert
                expect(underTest).toThrowErrorMatchingSnapshot();
            },
        );

        it.each([null, "not a function"])(
            "should throw if invalid getResourceLoader is provided",
            (badGetResourceLoader: any) => {
                // Arrange

                // Act
                const underTest = () =>
                    new JSDOMConfiguration(jest.fn(), badGetResourceLoader);

                // Assert
                expect(underTest).toThrowErrorMatchingSnapshot();
            },
        );

        it("should throw if invalid afterEnvSetup is provided", () => {
            // Arrange

            // Act
            const underTest = () =>
                new JSDOMConfiguration(
                    jest.fn(),
                    jest.fn(),
                    "not a function" as any,
                );

            // Assert
            expect(underTest).toThrowErrorMatchingInlineSnapshot(
                `"Must provide valid callback for after env setup or null"`,
            );
        });
    });

    describe("#getFileList", () => {
        it("should invoke method passed at construction", async () => {
            // Arrange
            const fakeGetFileList = jest
                .fn()
                .mockReturnValue(Promise.resolve("FILE_LIST" as any));
            const underTest = new JSDOMConfiguration(
                fakeGetFileList,
                jest.fn(),
            );
            const fakeFetchFn = jest.fn();
            const fakeRenderAPI: any = "FAKE_RENDER_API";

            // Act
            const result = await underTest.getFileList(
                "URL",
                fakeRenderAPI,
                fakeFetchFn,
            );

            // Assert
            expect(fakeGetFileList).toHaveBeenCalledWith(
                "URL",
                fakeRenderAPI,
                fakeFetchFn,
            );
            expect(result).toBe("FILE_LIST");
        });
    });

    describe("#getResourceLoader", () => {
        it("should invoke method passed at construction", () => {
            // Arrange
            const fakeGetResourceLoader = jest
                .fn()
                .mockReturnValue("RESOURCE_LOADER" as any);
            const underTest = new JSDOMConfiguration(
                jest.fn(),
                fakeGetResourceLoader,
            );
            const fakeRenderAPI: any = "FAKE_RENDER_API";

            // Act
            const result = underTest.getResourceLoader("URL", fakeRenderAPI);

            // Assert
            expect(fakeGetResourceLoader).toHaveBeenCalledWith(
                "URL",
                fakeRenderAPI,
            );
            expect(result).toBe("RESOURCE_LOADER");
        });
    });

    describe("#afterEnvSetup", () => {
        it("should invoke method passed at construction", async () => {
            // Arrange
            const fakeAfterEnvSetup = jest.fn().mockResolvedValue(null);
            const underTest = new JSDOMConfiguration(
                jest.fn(),
                jest.fn(),
                fakeAfterEnvSetup,
            );
            const fakeRenderAPI: any = "FAKE_RENDER_API";

            // Act
            await underTest.afterEnvSetup("URL", ["A", "B"], fakeRenderAPI);

            // Assert
            expect(fakeAfterEnvSetup).toHaveBeenCalledWith(
                "URL",
                ["A", "B"],
                fakeRenderAPI,
            );
        });

        it("should resolve to null if no method passed at construction", async () => {
            // Arrange
            const underTest = new JSDOMConfiguration(jest.fn(), jest.fn());
            const fakeRenderAPI: any = "FAKE_RENDER_API";

            // Act
            const result = await underTest.afterEnvSetup(
                "URL",
                ["A", "B"],
                fakeRenderAPI,
            );

            // Assert
            expect(result).toBeNull();
        });
    });
});
