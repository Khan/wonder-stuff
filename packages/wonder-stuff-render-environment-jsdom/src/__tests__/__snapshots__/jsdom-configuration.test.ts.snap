// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSDOMConfiguration #constructor should throw if invalid getFileList is provided 1`] = `"Must provide valid callback for obtaining file list"`;

exports[`JSDOMConfiguration #constructor should throw if invalid getFileList is provided 2`] = `"Must provide valid callback for obtaining file list"`;

exports[`JSDOMConfiguration #constructor should throw if invalid getResourceLoader is provided 1`] = `"Must provide valid callback for obtaining resource loader"`;

exports[`JSDOMConfiguration #constructor should throw if invalid getResourceLoader is provided 2`] = `"Must provide valid callback for obtaining resource loader"`;
